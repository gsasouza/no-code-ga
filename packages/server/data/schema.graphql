"""Algorithm data"""
type Algorithm implements Node {
  """The ID of an object"""
  id: ID!
  name: String
  user: User
  setup: Setup
}

"""A connection to a list of items."""
type AlgorithmConnection {
  """Information to aid in pagination."""
  pageInfo: PageInfo!

  """A list of edges."""
  edges: [AlgorithmEdge]
  count: Int
}

"""An edge in a connection."""
type AlgorithmEdge {
  """The item at the end of the edge"""
  node: Algorithm

  """A cursor for use in pagination"""
  cursor: String!
}

input AuthUserInput {
  token: String!
  clientMutationId: String
}

type AuthUserPayload {
  me: User
  token: String
  error: String
  clientMutationId: String
}

type DataModel {
  name: String
  type: String
}

type Mutation {
  UserAuth(input: AuthUserInput!): AuthUserPayload
}

"""An object with an ID"""
interface Node {
  """The id of the object."""
  id: ID!
}

"""Information about pagination in a connection."""
type PageInfo {
  """When paginating forwards, are there more items?"""
  hasNextPage: Boolean!

  """When paginating backwards, are there more items?"""
  hasPreviousPage: Boolean!

  """When paginating backwards, the cursor to continue."""
  startCursor: String

  """When paginating forwards, the cursor to continue."""
  endCursor: String
}

"""The root of all... queries"""
type Query {
  """Fetches an object given its ID"""
  node(
    """The ID of an object"""
    id: ID!
  ): Node
  me: User
  users(after: String, first: Int, before: String, last: Int, search: String): UserConnection
  algorithms(after: String, first: Int, before: String, last: Int, search: String): AlgorithmConnection
}

"""Setup data"""
type Setup {
  populationSize: Int
  generateFunction: String
  testFunction: String
  dataModel: [DataModel]
}

"""User data"""
type User implements Node {
  """The ID of an object"""
  id: ID!
  name: String
  email: String
  profilePicture: String
}

"""A connection to a list of items."""
type UserConnection {
  """Information to aid in pagination."""
  pageInfo: PageInfo!

  """A list of edges."""
  edges: [UserEdge]
  count: Int
}

"""An edge in a connection."""
type UserEdge {
  """The item at the end of the edge"""
  node: User

  """A cursor for use in pagination"""
  cursor: String!
}
