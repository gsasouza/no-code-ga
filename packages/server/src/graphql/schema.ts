import { GraphQLSchema } from 'graphql';

import MutationType from './type/MutationType';
import QueryType from './type/QueryType';

const schema = new GraphQLSchema({
  query: QueryType,
  mutation: MutationType,
});

export default schema;
