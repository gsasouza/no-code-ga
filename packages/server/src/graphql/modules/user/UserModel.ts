import mongoose from 'mongoose';

export interface IUser extends Document {
  name: string;
  email: string;
  picture: string;
}

const userSchema = new mongoose.Schema(
  {
    name: {
      type: String,
      required: true,
    },
    email: {
      type: String,
      required: true,
      trim: true,
    },
    picture: {
      type: String,
      required: false,
      trim: true,
    },
  },
  { timestamps: true },
);

export const User = conn => conn.model('User', userSchema);

export default User;
