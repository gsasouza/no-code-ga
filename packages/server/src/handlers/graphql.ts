import { ApolloServer } from 'apollo-server-lambda';
import { applyMiddleware } from 'graphql-middleware';

import { MONGO_URL, getConnection } from '../common';
import { dataloadersMiddleware } from '../graphql/loaders';
import schema from '../graphql/schema';

const server = new ApolloServer({
  schema: applyMiddleware(schema, dataloadersMiddleware),
  context: async () => {
    const dbConnection = await getConnection(MONGO_URL);
    return { dbConnection };
  },
  playground: {
    endpoint: '/dev/graphql',
  },
  introspection: true,
});

exports.handler = server.createHandler({
  cors: {
    origin: '*',
    credentials: true,
  },
});
