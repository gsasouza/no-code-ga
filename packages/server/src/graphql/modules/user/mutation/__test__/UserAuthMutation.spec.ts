import axios from 'axios';
import { graphql } from 'graphql';

import {
  clearDbAndRestartCounters,
  connectMongoose,
  disconnectMongoose,
  createUser,
  getContext,
} from '../../../../../../test/helpers';
import schema from '../../../../schema';

import { User } from '../../UserModel';

jest.mock('axios');

beforeAll(connectMongoose);

beforeEach(clearDbAndRestartCounters);

afterAll(disconnectMongoose);

describe('Auth User', () => {
  it('should add user to database if it doesnt exist ', async () => {
    axios.get.mockResolvedValue({
      data: {
        email: 'email@domain.com',
        name: 'name',
        picture: 'picture',
      },
    });
    // language=GraphQL
    const query = `
      mutation M($token: String!) {
        UserAuth(input: { token: $token }) {
          error
          token
          me {
            name
            email
            profilePicture
          }
        }
      }
    `;

    const variables = {
      token: 'VALID_TOKEN',
    };

    const rootValue = {};
    const context = await getContext();
    const result = await graphql(schema, query, rootValue, context, variables);

    expect(result.data).toMatchSnapshot();
    const users = await User(context.dbConnection).find({});

    expect(users.length).toBe(1);
  });

  it('should not add user to database if it exists ', async () => {
    await createUser({ email: 'email@domain.com' });

    axios.get.mockResolvedValue({
      data: {
        email: 'email@domain.com',
        name: 'name',
        picture: 'picture',
      },
    });
    // language=GraphQL
    const query = `
      mutation M($token: String!) {
        UserAuth(input: { token: $token }) {
          error
          token
          me {
            name
            email
            profilePicture
          }
        }
      }
    `;

    const variables = {
      token: 'VALID_TOKEN',
    };

    const rootValue = {};
    const context = await getContext();
    const result = await graphql(schema, query, rootValue, context, variables);

    expect(result.data).toMatchSnapshot();
    const users = await User(context.dbConnection).find({});
    expect(users.length).toBe(1);
  });

  it('should not auth an user with an invalid token', async () => {
    axios.get.mockResolvedValue(() => {
      throw 'Invalid token';
    });
    // language=GraphQL
    const query = `
      mutation M($token: String!) {
        UserAuth(input: { token: $token }) {
          error
          token
          me {
            name
            email
            profilePicture
          }
        }
      }
    `;

    const variables = {
      token: 'INVALID_TOKEN',
    };

    const rootValue = {};
    const context = await getContext();
    const result = await graphql(schema, query, rootValue, context, variables);

    expect(result.data).toMatchSnapshot();
    const users = await User(context.dbConnection).find({});
    expect(users.length).toBe(0);
  });
});
