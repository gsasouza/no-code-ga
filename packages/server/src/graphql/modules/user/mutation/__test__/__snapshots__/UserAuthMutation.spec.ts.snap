// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Auth User should add user to database if it doesnt exist  1`] = `
Object {
  "UserAuth": Object {
    "error": null,
    "me": Object {
      "email": "email@domain.com",
      "name": "name",
      "profilePicture": "picture",
    },
    "token": "VALID_TOKEN",
  },
}
`;

exports[`Auth User should not add user to database if it exists  1`] = `
Object {
  "UserAuth": Object {
    "error": null,
    "me": Object {
      "email": "email@domain.com",
      "name": "User#1",
      "profilePicture": null,
      "role": "user",
    },
    "token": "VALID_TOKEN",
  },
}
`;

exports[`Auth User should not auth an user with an invalid token 1`] = `
Object {
  "UserAuth": Object {
    "error": "Invalid token",
    "me": null,
    "token": null,
  },
}
`;
